File: src/stores/favorites.ts
import { defineStore } from "pinia";
import { ref } from "vue";
import { commands } from "@/bindings";
import type { Favorite } from "@/bindings";

export type { Favorite };

export const useFavoritesStore = defineStore("favorites", () => {
  const favorites = ref<Favorite[]>([]);
  const loading = ref(false);
  const error = ref<string | null>(null);

  async function loadFavorites() {
    loading.value = true;
    error.value = null;
    try {
      const result = await commands.getFavorites();
      if (result.status === "error") throw result.error;
      favorites.value = result.data;
    } catch (e) {
      error.value = String(e);
      console.error("Failed to load favorites:", e);
    } finally {
      loading.value = false;
    }
  }

  async function addFavorite(entity: {
    id: string;
    name: string;
    kind: string;
    slug: string;
    uuid?: string;
  }) {
    try {
      const result = await commands.addFavorite(
        entity.id,
        entity.name,
        entity.kind,
        entity.slug,
        entity.uuid ?? "",
      );
      if (result.status === "error") throw result.error;
      await loadFavorites();
    } catch (e) {
      console.error("Failed to add favorite:", e);
      throw e;
    }
  }

  async function removeFavorite(id: string, kind: string) {
    try {
      const result = await commands.removeFavorite(id, kind);
      if (result.status === "error") throw result.error;
      await loadFavorites();
    } catch (e) {
      console.error("Failed to remove favorite:", e);
      throw e;
    }
  }

  function isFavorite(id: string, kind: string): boolean {
    return favorites.value.some((f) => f.id === id && f.kind === kind);
  }

  async function toggleFavorite(entity: {
    id: string;
    name: string;
    kind: string;
    slug: string;
    uuid?: string;
  }) {
    if (isFavorite(entity.id, entity.kind)) {
      await removeFavorite(entity.id, entity.kind);
    } else {
      await addFavorite(entity);
    }
  }

  return {
    favorites,
    loading,
    error,
    loadFavorites,
    addFavorite,
    removeFavorite,
    isFavorite,
    toggleFavorite,
  };
});

File: src-tauri/src/commands/favorites.rs
use log::info;
use serde::{Deserialize, Serialize};
use specta::Type;
use tauri::State;

use crate::state::AppState;

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
pub struct Favorite {
    pub id: String,
    pub name: String,
    pub kind: String,
    pub slug: String,
    pub uuid: String,
    pub source: String,
    pub added_at: String,
}

#[tauri::command]
#[specta::specta]
pub async fn get_favorites(state: State<'_, AppState>) -> Result<Vec<Favorite>, String> {
    let db = state.cache.db().lock().unwrap();
    let mut stmt = db
        .prepare("SELECT id, name, kind, slug, uuid, source, added_at FROM favorites ORDER BY kind, name")
        .map_err(|e| format!("Failed to prepare favorites query: {}", e))?;

    let rows = stmt
        .query_map([], |row| {
            Ok(Favorite {
                id: row.get(0)?,
                name: row.get(1)?,
                kind: row.get(2)?,
                slug: row.get(3)?,
                uuid: row.get(4)?,
                source: row.get(5)?,
                added_at: row.get(6)?,
            })
        })
        .map_err(|e| format!("Failed to query favorites: {}", e))?;

    let mut favorites = Vec::new();
    for row in rows {
        favorites.push(row.map_err(|e| format!("Failed to read favorite row: {}", e))?);
    }
    Ok(favorites)
}

#[tauri::command]
#[specta::specta]
pub async fn add_favorite(
    id: String,
    name: String,
    kind: String,
    slug: String,
    uuid: String,
    source: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let db = state.cache.db().lock().unwrap();
    db.execute(
        "INSERT OR IGNORE INTO favorites (id, name, kind, slug, uuid, source) VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
        rusqlite::params![id, name, kind, slug, uuid, source],
    )
    .map_err(|e| format!("Failed to add favorite: {}", e))?;

    info!("Added favorite: {} ({}, source={})", name, kind, source);
    Ok(())
}

#[tauri::command]
#[specta::specta]
pub async fn remove_favorite(
    id: String,
    kind: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let db = state.cache.db().lock().unwrap();
    db.execute(
        "DELETE FROM favorites WHERE id = ?1 AND kind = ?2",
        rusqlite::params![id, kind],
    )
    .map_err(|e| format!("Failed to remove favorite: {}", e))?;

    info!("Removed favorite: {} ({})", id, kind);
    Ok(())
}

#[tauri::command]
#[specta::specta]
pub async fn is_favorite(
    id: String,
    kind: String,
    state: State<'_, AppState>,
) -> Result<bool, String> {
    let db = state.cache.db().lock().unwrap();
    let count: i64 = db
        .query_row(
            "SELECT COUNT(*) FROM favorites WHERE id = ?1 AND kind = ?2",
            rusqlite::params![id, kind],
            |row| row.get(0),
        )
        .map_err(|e| format!("Failed to check favorite: {}", e))?;
    Ok(count > 0)
}
